spec/app/ics-031-crosschain-queries/README.md

```diff
@@ -0,0 +1,388 @@
+            proof,
+            query.path,
+        ))
+        result = FAILURE
+    }
+
+    // Delete the query from the local, private store.
+    privateStore.delete(queryPath(queryId))
+
+    // Create a query result record.
+    resultRecord = CrossChainQuery{queryIdentifier,
+                                   result,
+                                   data} 
+
+    // Store the result in a public path.
+    provableStore.set(resultQueryPath(queryIdentifier), resultRecord)
+
+}
+```
+- **Precondition**
+  - There exists a client with `clientId` identifier.
+  - There is a query request stored in the `privateStore` identified by `queryId`.
+- **Postcondition**
+  - The query request identified by `queryId` is deleted from the `privateStore`.
+  - The query result is stored in the `provableStore`.
+
+The `PruneCrossChainQueryResult` function is called when the caller of a query has retrieved the result and wants to delete it.
+
+```typescript
+function PruneCrossChainQueryResult(
+  queryId: Identifier,
+  queryCapability: CapabilityKey
+  ) {
+
+    // Retrieve the query result from the provable store using the query's identifier.
+    resultRecord = privateStore.get(resultQueryPath(queryIdentifier))
+    abortTransactionUnless(resultRecord !== null)
+
+    // Abort the transaction unless the caller has the right to clean the query result
+    abortTransactionUnless(authenticateCapability(queryId, queryCapability))
+
+    // Delete the query result from the public store.
+    privateStore.delete(resultQueryPath(queryId))
+}
+```
+- **Precondition**
+  - There is a query result stored in the `provableStore` identified by `queryId`.
+  - The caller has the right to clean the query result
+- **Postcondition**
+  - The query result identified by `queryId` is deleted from the `provableStore`.
+
+#### Timeouts
+
+Query requests have associated a `localTimeoutHeight` and a `localTimeoutTimestamp` field that specifies the height and timestamp limit at the Querying Chain after which a query is considered to have failed. 
+
+The Querying Chain calls the `checkQueryTimeout` function to check whether a specific query has timed out. 
+
+> There are several alternatives on how to handle timeouts. For instance, the relayer could submit on-chain timeout notifications to the Querying Chain. Since the relayer is untrusted, for each of these notifications the Cross-chain Querying module of the Querying Chain MUST call the `checkQueryTimeout` to check if the query has indeed timed out. An alternative could be to make the Cross-chain Querying module responsible for checking  
+if any query has timed out by iterating over the ongoing queries at the beginning of a block and calling `checkQueryTimeout`. This is an implementation detail that this specification does not cover.
+
+We pass the relayer address just as in `CrossChainQueryResult` to allow for possible incentivization here as well.
+
+```typescript
+function checkQueryTimeout(
```

Comment:
We should store state in such a way that we can efficiently iterate over existing queries in timeout order (for timestamp and height), though this may be left as details for implementation

Reply:
I think this is an implementation detail. I have added the following sentence: "In this case, ongoing queries should be stored indexed by `localTimeoutTimestamp` and `localTimeoutHeight` to allow iterating over them more efficiently"